mrmustard/lab_dev/circuit_components.py

```diff
@@ -48,16 +48,35 @@ class CircuitComponent:
     def __init__(
         self,
         name: Optional[str] = None,
-        representation: Optional[Representation] = None,
-        modes_out_bra: Optional[Sequence[int]] = None,
-        modes_in_bra: Optional[Sequence[int]] = None,
-        modes_out_ket: Optional[Sequence[int]] = None,
-        modes_in_ket: Optional[Sequence[int]] = None,
+        representation: Optional[Bargmann | Fock] = None,
+        modes_out_bra: Optional[Sequence[int]] = (),
+        modes_in_bra: Optional[Sequence[int]] = (),
+        modes_out_ket: Optional[Sequence[int]] = (),
+        modes_in_ket: Optional[Sequence[int]] = (),
     ) -> None:
-        self._name = name or ""
-        self._wires = Wires(modes_out_bra, modes_in_bra, modes_out_ket, modes_in_ket)
+        self._wires = Wires(
+            set(modes_out_bra), set(modes_in_bra), set(modes_out_ket), set(modes_in_ket)
+        )
+        self._name = name or "CC" + "".join(str(m) for m in sorted(self.wires.modes))
         self._parameter_set = ParameterSet()
         self._representation = representation
+        # handle out-of-order modes
+        a, b, c, d = (
+            sorted(modes_out_bra),
+            sorted(modes_in_bra),
+            sorted(modes_out_ket),
+            sorted(modes_in_ket),
+        )
+        if a != sorted(a) or b != sorted(b) or c != sorted(c) or d != sorted(d):
+            offsets = [0, len(a), len(a) + len(b), len(a) + len(b) + len(c)]
+            perm = (
+                tuple(np.argsort(a))
+                + tuple(np.argsort(b) + offsets[0])
+                + tuple(np.argsort(c) + offsets[1])
+                + tuple(np.argsort(d) + offsets[2])
+            )
+            if self._representation is not None:
+                self._representation = self._representation.reorder(tuple(perm))
```

Comment:
:+1:  pretty cool
Just a reminder that we will need to write the same piece of code to reorder the parameter set